fuel-txpool/src/containers/dependency.rs

```diff
@@ -435,75 +439,72 @@ impl Dependency {
         Ok(removed_tx)
     }
 
-    pub fn recursively_remove_all_dependencies<'a>(
+    /// Remove all pending txs that depend on the outputs of the provided tx
+    pub(crate) fn recursively_remove_all_dependencies<'a>(
         &'a mut self,
         txs: &'a HashMap<TxId, TxInfo>,
         tx: ArcTx,
     ) -> Vec<ArcTx> {
-        let mut removed_tx = vec![tx.clone()];
+        let mut removed_transactions = vec![tx.clone()];
+        let id = tx.id();
 
-        // for all outputs recursivly call removal
+        // recursively remove all transactions dependant on the outputs of the current tx
         for (index, output) in tx.outputs().iter().enumerate() {
-            // get contract_id or none if it is coin
-            let is_contract = match output {
-                Output::Coin { .. }
-                | Output::Withdrawal { .. }
-                | Output::Change { .. }
-                | Output::Variable { .. } => None,
-                Output::Contract { input_index, .. } => {
-                    if let Input::Contract { contract_id, .. } = tx.inputs()[*input_index as usize]
-                    {
-                        Some(contract_id)
+            match output {
+                Output::Withdrawal { .. } | Output::Contract { .. } => {
+                    // no other transactions can depend on these types of outputs
+                }
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    // remove transactions that depend on this coin output
+                    let utxo = UtxoId::new(tx.id(), index as u8);
+                    if let Some(state) = self.coins.remove(&utxo).map(|c| c.is_spend_by) {
+                        // there may or may not be any dependents for this coin output
+                        if let Some(spend_by) = state {
+                            let rem_tx = txs.get(&spend_by).expect("Tx should be present in txs");
+                            removed_transactions.extend(
+                                self.recursively_remove_all_dependencies(txs, rem_tx.tx().clone()),
+                            );
+                        }
                     } else {
-                        panic!("InputIndex for contract should be always correct");
```

Comment:
I prefere to panic here

Reply:
I think this case isn't so egregious that we need to crash the node. I'd rather come up with a way to reset the mempool if it's unhappy about its' own state management than to panic and crash the rest of the node.